File: autocontext/src/autocontext/knowledge/pending_lessons.py
"""Per-scenario queue of lessons awaiting human approval (Cowork 2c)."""

from __future__ import annotations

import json
import logging
from collections.abc import Sequence
from pathlib import Path

from autocontext.knowledge.lessons import Lesson

logger = logging.getLogger(__name__)


class PendingLessonStore:
    """JSON-backed queue of pending lessons, mirroring LessonStore's on-disk shape."""

    def __init__(self, knowledge_root: Path) -> None:
        self.knowledge_root = knowledge_root

    def _path(self, scenario: str) -> Path:
        from autocontext.storage.scenario_paths import resolve_scenario_root

        return resolve_scenario_root(self.knowledge_root, scenario) / "pending_lessons.json"

    def read(self, scenario: str) -> list[Lesson]:
        path = self._path(scenario)
        if not path.exists():
            return []
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            if not isinstance(data, list):
                return []
        except (OSError, TypeError, ValueError, json.JSONDecodeError):
            logger.debug("unable to read pending lessons for %s from %s", scenario, path)
            return []
        return [Lesson.from_dict(entry) for entry in data]

    def write(self, scenario: str, lessons: Sequence[Lesson]) -> None:
        path = self._path(scenario)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(
            json.dumps([les.to_dict() for les in lessons], indent=2),
            encoding="utf-8",
        )

    def add(self, scenario: str, lesson: Lesson) -> None:
        lessons = self.read(scenario)
        if any(existing.id == lesson.id or existing.text == lesson.text for existing in lessons):
            return
        lessons.append(lesson)
        self.write(scenario, lessons)

    def remove(self, scenario: str, lesson_id: str) -> Lesson | None:
        lessons = self.read(scenario)
        removed: Lesson | None = None
        kept: list[Lesson] = []
        for lesson in lessons:
            if lesson.id == lesson_id and removed is None:
                removed = lesson
            else:
                kept.append(lesson)
        if removed is not None:
            self.write(scenario, kept)
        return removed

File: ts/src/knowledge/pending-lessons.ts
/** Per-scenario queue of lessons awaiting human approval (Cowork 2c). */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { type Lesson, lessonFromDict, lessonToDict } from "./lessons.js";

export class PendingLessonStore {
  constructor(private readonly knowledgeRoot: string) {}

  private path(scenario: string): string {
    return join(this.knowledgeRoot, scenario, "pending_lessons.json");
  }

  read(scenario: string): Lesson[] {
    const p = this.path(scenario);
    if (!existsSync(p)) return [];
    try {
      const data = JSON.parse(readFileSync(p, "utf-8"));
      if (!Array.isArray(data)) return [];
      return data.map((e) => lessonFromDict(e as Record<string, unknown>));
    } catch {
      return [];
    }
  }

  write(scenario: string, lessons: Lesson[]): void {
    mkdirSync(join(this.knowledgeRoot, scenario), { recursive: true });
    writeFileSync(this.path(scenario), JSON.stringify(lessons.map(lessonToDict), null, 2), "utf-8");
  }

  add(scenario: string, lesson: Lesson): void {
    const lessons = this.read(scenario);
    if (lessons.some((e) => e.id === lesson.id || e.text === lesson.text)) return;
    lessons.push(lesson);
    this.write(scenario, lessons);
  }

  remove(scenario: string, lessonId: string): Lesson | null {
    const lessons = this.read(scenario);
    let removed: Lesson | null = null;
    const kept: Lesson[] = [];
    for (const l of lessons) {
      if (l.id === lessonId && removed === null) removed = l;
      else kept.push(l);
    }
    if (removed !== null) this.write(scenario, kept);
    return removed;
  }
}
